src/components/Settings/Settings.tsx

```diff
@@ -235,49 +123,48 @@ export class Settings extends Component<Props, State> {
      * @param resourceType  ResourceTypes, the type of the control that changed
      */
     handleToggleChange = (isOn: boolean, resourceType: ResourceTypes) => {
-        let { cameraOn, audioOn } = this.state;
+        const { cameraOn, audioOn } = this.state;
+
+        let newCameraOn = resourceType === ResourceTypes.VIDEO ? isOn : cameraOn;
+        let newAudioOn = resourceType === ResourceTypes.AUDIO ? isOn : audioOn;
 
         // do not allow both camera and audio to be turned off
-        cameraOn = !isOn && resourceType === ResourceTypes.AUDIO ? true : cameraOn;
-        audioOn = !isOn && resourceType === ResourceTypes.VIDEO ? true : audioOn;
+        newCameraOn = !isOn && resourceType === ResourceTypes.AUDIO ? true : newCameraOn;
+        newAudioOn = !isOn && resourceType === ResourceTypes.VIDEO ? true : newAudioOn;
 
-        this.setState(
-            (prevState: State) => {
-                return {
-                    cameraOn: resourceType === ResourceTypes.VIDEO ? isOn : cameraOn,
-                    audioOn: resourceType === ResourceTypes.AUDIO ? isOn : audioOn,
-                    screenOn:
-                        resourceType === ResourceTypes.SCREEN_SHARE ? isOn : prevState.screenOn
-                };
-            },
-            () => {
-                this.saveSettings();
-            }
-        );
-    };
+        this.setState((prevState: State) => {
+            return {
+                cameraOn: newCameraOn,
+                audioOn: newAudioOn,
+                screenOn: resourceType === ResourceTypes.SCREEN_SHARE ? isOn : prevState.screenOn
+            };
+        });
 
-    saveSettings = () => {
         if (this.props.onSettingsChanged) {
-            const camera = this.state.cameraOn ? this.state.selectedCamera : false;
-            const audio = this.state.audioOn ? this.state.selectedAudio : false;
+            const camera = newCameraOn
+                ? this.props.selectedCameraDevice || this.cameraDevicesInfo[0]
+                : false;
+            const audio = newAudioOn
+                ? this.props.selectedAudioDevice || this.audioDevicesInfo[0]
+                : false;
             this.props.onSettingsChanged(camera, audio, this.state.screenOn);
```

Comment:
please avoid that kind of hard-to-read conditions and split the entire logic into small pieces:
```
let { onSettingsChanged, selectedCameraDevice, selectedAudioDevice, onSettingsChanged } = this.props;
if(!onSettingsChanged){
  return;
}
if(cameraOn){
  selectedCameraDevice = selectedCameraDevice || this.cameraDevicesInfo[0];
}
...

onSettingsChanged(selectedCameraDevice, selectedAudioDevice, this.state.screenOn);
```

Reply:
I find ? opertaor more readable then nested conditions, but I'' add variable decleration